url_finder/src/find_url.rs

```diff
@@ -0,0 +1,274 @@
+        .address
+        .strip_prefix("f0")
+        .unwrap_or(&payload.address)
+        .to_string();
+
+    // Find piece_cid from deals and test if the url is working, returning the first working url
+    let working_url = get_working_url(state, endpoints, &provider, payload.extended_search)
+        .await
+        .map_err(|e| {
+            debug!("Failed to get working url: {:?}", e);
+            bad_request("Failed to get working url")
+        })?;
+
+    Ok(ok_response(FindUrlResponse { url: working_url }))
+}
+
+async fn get_peer_id_from_lotus_rpc(address: &str) -> Result<String, Error> {
+    let client = Client::new();
+    let res = client
+        .post("https://api.node.glif.io/rpc/v1")
+        .json(&json!({
+            "jsonrpc": "2.0",
+            "id": 1,
+            "method": "Filecoin.StateMinerInfo",
+            "params": [address, null]
+        }))
+        .send()
+        .await?
+        .json::<serde_json::Value>()
+        .await?;
+
+    let peer_id = res
+        .get("result")
+        .unwrap()
+        .get("PeerId")
+        .unwrap()
+        .as_str()
+        .unwrap();
+
+    Ok(peer_id.to_string())
+}
+
+async fn get_res_from_cid_contant(peer_id: &str) -> Result<serde_json::Value> {
+    let client = Client::new();
+    let url = format!("https://cid.contact/providers/{}", peer_id);
+
+    debug!("cid contact url: {:?}", url);
+
+    let res = client
+        .get(&url)
+        .send()
+        .await?
+        .json::<serde_json::Value>()
+        .await?;
+
+    Ok(res)
+}
+
+fn get_all_addresses_from_cid_contact_response(json: serde_json::Value) -> Result<Vec<String>> {
+    let mut addresses = vec![];
+
+    // let keys = ["AddrInfo", "Publisher"];
+    // for key in &keys {
+    //     if let Some(addrs_arr) = json
+    //         .get(*key)
+    //         .and_then(|item| item.get("Addrs"))
+    //         .and_then(|addrs| addrs.as_array())
+    //     {
+    //         addrs_arr
+    //             .iter()
+    //             .filter_map(|addr| addr.as_str())
+    //             .for_each(|addr_str| addresses.push(addr_str.to_string()));
+    //     }
+    // }
+
+    if let Some(e_providers) = json
+        .get("ExtendedProviders")
+        .and_then(|ep| ep.get("Providers"))
+        .and_then(|p| p.as_array())
+    {
+        e_providers
+            .iter()
+            .filter_map(|provider| provider.get("Addrs"))
+            .filter_map(|addrs| addrs.as_array())
+            .flat_map(|addrs_arr| addrs_arr.iter())
+            .filter_map(|addr| addr.as_str())
+            .for_each(|addr| {
+                addresses.push(addr.to_string());
+            });
+    }
+
+    if addresses.is_empty() {
+        bail!("No addresses found");
+    }
+
+    Ok(addresses)
+}
+
+fn parse_addrs_to_endpoints(addrs: Vec<String>) -> Result<Vec<String>, Error> {
+    let mut endpoints = vec![];
+
+    for addr in addrs {
+        let parts: Vec<&str> = addr.split("/").collect();
```

Comment:
Technically `addrs` is a Multiaddr: https://multiformats.io/multiaddr/ . Let's use a dedicated crate for parsing: https://github.com/multiformats/rust-multiaddr

Reply:
not rly helpful

Reply:
Fair enough :+1: